### DISCUSSION TITLE
Dusting of images

### DISCUSSION THREAD

**Discussion by dhegedus99** [seung-lab/connected-components-3d#107] [Q&A] [answered]
I'm using this package to find spatiotemporally contiguous heatwaves in remotely-sensed land surface temperature data (MODIS-LST). I have created a binary 3D (latitude, longitude, time) array, where heawave days are 1, and non-heatwave days are 0. Due to the nature of the dataset, it is quite noisy, so the individual heatwaves aren't always connected. The connected_components method finds way too many small heatwaves (which are actually part of a bigger heatwave). However, when I try dusting the array to get rid of the small clusters, I get a "TypeError: No matching signature found" if I choose a threshold other than 1 (which does not change the array at all). Do you have any idea why the dusting might not work for me with a threshold other than 1? Any suggestions how to solve this problem?
Thanks!

**Comment by william-silversmith** [comment]
Hi! Thanks for using cc3d! I'm very happy you're finding interesting uses for the software.

I suspect that you may be using an old buggy version of dust. Can you let me know what OS, Python version, and cc3d version you're using? I tried running dust on a binary noise array with types uint8 and bool and didn't trigger this error. Note that Python3.6 and below are EOL and do not download current versions of cc3d from PyPI. Upgrade your Python version in that case.

**Reply by dhegedus99** [reply]
Thanks for the quick reply!
I installed cc3d (version 3.10.3) with pip install connected-components-3d - maybe that's not the latest version? The python environment has 3.8.15 python. I'm running this code in a Jupyter Notebook (through Anaconda) on a Windows OS.

**Reply by william-silversmith** [reply]
Okay, that's really weird. Can you give me a code snippet that reproduces your problem? You can use [`np.random.randint`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html) to make some fake data.

**Reply by dhegedus99** [reply]
<img width="602" alt="code_snippet_dusting" src="https://user-images.githubusercontent.com/78150129/217240171-af40b7cc-5bb5-4c31-a179-3c5e5bf76a8c.png">
I've tried with some fake data, and it still gives me the same error, I've attached a screenshot of the error message. Here's the code:
`fake_data = np.random.randint(2, size=(2116,20,360))
labels_out = cc3d.dust(fake_data, threshold=10, connectivity=26, in_place=False)`

**Reply by william-silversmith** [reply]
Thank you! I can reproduce this now. Will look into this.

**Reply by william-silversmith** [reply]
Interesting... this only occurs when the test data are large.

**Reply by william-silversmith** [reply]
Here's one workaround: `fake_data = np.random.randint(2, size=(2116,20,360), dtype=bool)` This will reduce the size of the data 8x (it was previously int64) and seems to work. I will have to investigate more to find out the root cause.

**Reply by dhegedus99** [reply]
Thank you so much! That works for me well!

**Reply by william-silversmith** [reply]
I figured it out. I wrote parts of the dust function to only accept unsigned integers. I should fix this, but stick to uints and you should be fine.

**Comment by william-silversmith** [comment, marked as answer]
As noted in the previous discussion, dust currently only accepts unsigned ints in 3.10.4. I'll be releasing 3.10.5 which expands the acceptable types to signed ints.
